### DISCUSSION TITLE
Using directly IOCP Windows

### DISCUSSION THREAD

**Discussion by ghost** [curl/curl#8175] [Q&A]
Hello,

Let me ask please: I am looking at an example libcurl: https://curl.se/libcurl/c/externalsocket.html

There is a line of code like this:

```
/* Create the socket "manually" */
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
```

I unfortunately cannot understand, is this socket created by an operating system function (for example, windows) or by a Сurl function?
Simply, if the socket is created by the operating system, then how can create an asynchronous Windows socket - WSASocket?
Would it be legal to pass the created WSASocket to the callback function - Opensocket?

**Comment by bagder** [comment]
> I unfortunately cannot understand, is this socket created by an operating system function (for example, windows) or by a Сurl function?

That socket is created by your application, and handed to libcurl when it needs a socket for the transfer.

> Would it be legal to pass the created WSASocket to the callback function - Opensocket?

It think so. But this is an area where not too many others have traveled before...

**Comment by bagder** [comment]
The mention of I/O completion ports has been removed from everything curl.
